File: compiler/src/diagnostic.rs
/// Diagnostic messages produced during compilation.

pub enum DiagnosticLevel {
    Warning,
    Error,
}

pub struct Diagnostic {
    pub level: DiagnosticLevel,
    pub path: String,
    pub message: String,
    pub line: usize,
    pub column: usize,
}

impl Diagnostic {
    pub fn error(
        path: String,
        message: String,
        line: usize,
        column: usize,
    ) -> Self {
        Diagnostic {
            level: DiagnosticLevel::Error,
            path: path,
            message: message,
            line: line,
            column: column,
        }
    }

    pub fn warning(
        path: String,
        message: String,
        line: usize,
        column: usize,
    ) -> Self {
        Diagnostic {
            level: DiagnosticLevel::Warning,
            path: path,
            message: message,
            line: line,
            column: column,
        }
    }
}

File: compiler/lib/inkoc/diagnostic.rb
# frozen_string_literal: true

module Inkoc
  class Diagnostic
    attr_reader :level, :message, :location

    def self.error(*args)
      new(:error, *args)
    end

    def self.warning(*args)
      new(:warning, *args)
    end

    def initialize(level, message, location)
      @level = level
      @message = message
      @location = location
    end

    def error?
      @level == :error
    end

    def line
      location.line
    end

    def column
      location.column
    end

    def file
      location.file
    end

    def path
      file.path
    end
  end
end
